### PR TITLE
Added XSLT Injection Scanner

### PR DESCRIPTION
**Pull request by CaptainFreak**
Edit: See zaproxy/zaproxy#3572

### CODE REVIEW

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    														"\"/><xsl:value-of select=\"system-property(\'xsl:vendor\')\"/><!--",
+    														"<xsl:value-of select=\"system-property(\'xsl:vendor\')\"/><!--"
+    														};
+    private static final String[] xsltVendors={"libxslt","Microsoft","Saxonica","Apache","Xalan","SAXON","Transformiix"};
+    
+    //TODO Add time based payloads (SSRF)
+    
+	@Override
+	public int getId() {
+		return 90017;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: You'll need to do a separate PR to reserve the plugin ID.

https://github.com/zaproxy/zaproxy/blob/develop/src/doc/scanners.md

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+	// wasc_23 is XML Injection
+	private static Vulnerability vuln = Vulnerabilities.getVulnerability("wasc_23");
+    
+	// plugin dependencies
+    private static final String[] dependency = {};
+    
+	//For logging
+	private static Logger log = Logger.getLogger(XSLTInjection.class);
+    private boolean debugEnabled = log.isDebugEnabled();
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: What's the value in storing this locally? Can't `log.isDebugEnabled()` be called in the conditionals below?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+public class XSLTInjection extends AbstractAppParamPlugin {
+
+	// wasc_23 is XML Injection
+	private static Vulnerability vuln = Vulnerabilities.getVulnerability("wasc_23");
+    
+	// plugin dependencies
+    private static final String[] dependency = {};
+    
+	//For logging
+	private static Logger log = Logger.getLogger(XSLTInjection.class);
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Should be `final` and then changed to full caps `LOG` since it's a constant.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+/**
+ * The XSLTInjection plugin identifies XSLT injection vulnerabilities with all the parameters in GET and POST
+ *
+ * @author CaptainFreak
+ */
+public class XSLTInjection extends AbstractAppParamPlugin {
+
+	// wasc_23 is XML Injection
+	private static Vulnerability vuln = Vulnerabilities.getVulnerability("wasc_23");
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: `vuln` is unused and should be removed.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    
+	// plugin dependencies
+    private static final String[] dependency = {};
+    
+	//For logging
+	private static Logger log = Logger.getLogger(XSLTInjection.class);
+    private boolean debugEnabled = log.isDebugEnabled();
+
+    // Payloads
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: The comment and blank line can be removed, the arrays are named `___Payloads`

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    //TODO Add time based payloads (SSRF)
+    
+	@Override
+	public int getId() {
+		return 90017;
+	}
+
+	@Override
+	public String getName() {
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This blank line can be removed.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	@Override
+	public String getReference() {
+		return Constant.messages.getString("ascanalpha.xsltinjection.xslt.refs");
+	}
+
+	@Override
+	public void init() {
+        //DEBUG: turn on for debugging
+        //log.setLevel(org.apache.log4j.Level.DEBUG);
+        //this.debugEnabled = true;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: If `setLevel` is `DEBUG` then `isDebugEnabled()` returns true, no need to set two things...

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+        if (this.debugEnabled) {
+            log.debug("Initialising XSLT Scanner.");
+        }
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This blank can be removed

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 1 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: The ID should be registered in https://github.com/zaproxy/zaproxy/blob/develop/src/doc/scanners.md

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	public void scan(HttpMessage msg, String param, String value) {
+		try {
+			//Count positive Tests
+			int positiveBasedOnError=0; 
+			//error causing tests
+			for (int i=0;i<errorCausingPayloads.length;i++){
+				msg = getNewMsg();
+				setParameter(msg, param, errorCausingPayloads[i]);
+				sendAndReceive(msg);
+				if(!msg.getResponseBody().toString().contains(errorCausingPayloads[i])){
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Isn't this logic backward? Shouldn't the counter only be incremented if the payload is found?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+				}
+			}
+			//vendor returning tests
+			for(int i=0;i<vendorReturningPayloads.length;i++){
+				msg=getNewMsg();
+				setParameter(msg,param,vendorReturningPayloads[i]);
+				sendAndReceive(msg);
+				for(int j=0;j<xsltVendors.length;j++){
+					if(msg.getResponseBody().toString().contains(xsltVendors[j])){
+							if(positiveBasedOnError>0){
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Also the logic here means that there can only be an error based results when there's also a vendor result, was that intentional?

Shouldn't this be done in the previous loop? If an alert is raised there's no point continuing further checking efforts.

Or do you purposefully only alert if both are successful?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+								return;
+							}
+					}
+				}
+			}
+			
+			// TODO add time based testing and improve accuracy
+			
+			// Stop the testing if user requests
+			if (isStop()) {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This should be checked inside the loops.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	@Override
+	public int getRisk() {
+		return Alert.RISK_HIGH;
+	}
+
+	@Override
+	public int getCweId() {
+		// The CWE id
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Comment can be removed, I think the function name makes this clear.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+		return 91;
+	}
+
+	@Override
+	public int getWascId() {
+		// The WASC ID
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Comment can be removed, I think the function name makes this clear.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml [hunk 14]
```diff
@@ -34,6 +34,7 @@
 		<ascanrule>org.zaproxy.zap.extension.ascanrulesAlpha.TestUserAgent</ascanrule>
 		<ascanrule>org.zaproxy.zap.extension.ascanrulesAlpha.TraceAxdScanner</ascanrule>
 		<ascanrule>org.zaproxy.zap.extension.ascanrulesAlpha.HtAccessScanner</ascanrule>
+		<ascanrule>org.zaproxy.zap.extension.ascanrulesAlpha.XSLTInjection</ascanrule>
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml]: An entry should also be added to the changes section of the manifest.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 9 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: What are the expected contents of the response? Do you have a test page? (We could add that as unit test.)

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties [hunk 15]
```diff
@@ -145,3 +145,9 @@ ascanalpha.traceaxdscanner.otherinfo = Based on response status code Trace Viewe
 ascanalpha.traceaxdscanner.refs = https://msdn.microsoft.com/en-us/library/bb386420.aspx\nhttps://msdn.microsoft.com/en-us/library/wwh16c6c.aspx\nhttps://www.dotnetperls.com/trace
 ascanalpha.traceaxdscanner.soln = Consider whether or not Trace Viewer is actually required in production, if it isn't then disable it. If it is then ensure access to it requires authentication and authorization.
 
+ascanalpha.xsltinjection.xslt.name = XSLT Injection
+ascanalpha.xsltinjection.xslt.desc = Injection in XSLT files can result in multiple critical vulnerabilities. 
+ascanalpha.xsltinjection.xslt.refs = https://www.contextis.com/blog/xslt-server-side-injection-attacks
+ascanalpha.xsltinjection.xslt.soln = Sanitize and analyze every user input coming from any client-side.
+
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties]: One of these blanks can be removed.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 1 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: I'd suggest 40033 or 20020

Comment (thc202) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Is that for a special reason?

Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: No they just seemed like the next logical numbers. Honestly the groupings have largely been a mystery to me (I know there was, or is, some passive/active grouping, but outside of that I dunno). 90017 is fine too, I just figured there was no reason to be top-end-heavy..... <shrug>

Comment (thc202) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: OK. (Yeah, there was some groups/categories from Paros days but they were not followed always.)

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 9 above)
Comment (CaptainFreak) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: No. As in this case payloads are special characters of xslt and injecting them cause compile time errors and in realistic scenario (PHP7 + libxslt) like the one I used, It doesn't return error or anything. It just drops the whole element from DOM (also production apps never return errors). This tests will be improved over time after researching in various xslt processors and real life scenarios but its really obscure vuln and hence all the accuracy testing will be done on self written vuln sites. I will do it over the time.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 10 above)
Comment (CaptainFreak) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: yeah I did that purposefully because we will be more sure about the injection point plus it will give us a nice payload showing the context of injection and vendor.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 9 above)
Comment (CaptainFreak) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Expected responses include empty DOM element and vendor of xslt processor. Yes, I did setup a local implementation.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 1 above)
Comment (CaptainFreak) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: So 90017 is okay right ? I chose that because its near to injection vulnerabilities.

**Comment by CaptainFreak** [comment]
@kingthorin 
XSLT Injection demo : https://gist.github.com/CaptainFreak/bf8fb57f5029d29569f76bc7f67c5389

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 9 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Ok

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 10 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Ok

**Comment by CaptainFreak** [comment]
Any more suggestions ?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 1 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Yup

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+	public String getReference() {
+		return Constant.messages.getString("ascanalpha.xsltinjection.xslt.refs");
+	}
+
+	@Override
+	public void init() {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Initialising XSLT Scanner.");
+        }
+	}
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Indentation should still be normalized spaces or tabs. It's particularly noticeable via the github web interface.

https://github.com/zaproxy/zap-extensions/pull/1384/files

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+				msg = getNewMsg();
+				setParameter(msg, param, errorCausingPayloads[i]);
+				sendAndReceive(msg);
+				if(!msg.getResponseBody().toString().contains(errorCausingPayloads[i])){
+					positiveBasedOnError++;
+				}
+				// Stop the testing if user requests
+				if (isStop()) {
+					if (LOG.isDebugEnabled())
+						LOG.debug("Stopping the scan due to a user request");
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: While this is syntactically correct it would be nice if all the if blocks consistently used curly braces. (Throughout)

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml [hunk 18]
```diff
@@ -10,7 +10,7 @@
 	Add .htaccess scanner (Issue 3972).<br>
 	Modified trace.axd scanner to leverage new AbstractAppFilePlugin component.<br>
 	]]>
-	</changes>
+	<XSLTInjectionScanner>Added ascan rule for xslt injection vulnerabilities.</XSLTInjectionScanner></changes>
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml]: This should be inside the CDATA block above (eg: lines 10/11).

**Comment by CaptainFreak** [comment]
Done.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+/**
+ * The XSLTInjection plugin identifies XSLT injection vulnerabilities with all the parameters in GET and POST
+ *
+ * @author CaptainFreak
+ */
+public class XSLTInjection extends AbstractAppParamPlugin {
+	
+	// plugin dependencies
+	private static final String[] dependency = {};
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This (along with `getDependency`) can be removed, no need to override if there are no dependencies.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+	
+	//For logging
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This comment could be removed (the name/class already implies that).

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+	@Override
+	public void init() {
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Initialising XSLT Scanner.");
+		}
+	}
+
+	/*
+	 * This method is called by the active scanner for each GET and POST parameter for every page
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This comment can be removed.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+	/*
+	 * This method is called by the active scanner for each GET and POST parameter for every page 
+	 * @see org.parosproxy.paros.core.scanner.AbstractAppParamPlugin#scan(org.parosproxy.paros.network.HttpMessage, java.lang.String, java.lang.String)
+	 */
+	
+	@Override
+	public void scan(HttpMessage msg, String param, String value) {
+		try {
+			//Count positive Tests
+			int positiveBasedOnError=0;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Since the number of errors doesn't matter it should be used a boolean.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+	 * @see org.parosproxy.paros.core.scanner.AbstractAppParamPlugin#scan(org.parosproxy.paros.network.HttpMessage, java.lang.String, java.lang.String)
+	 */
+	
+	@Override
+	public void scan(HttpMessage msg, String param, String value) {
+		try {
+			//Count positive Tests
+			int positiveBasedOnError=0; 
+			//error causing tests
+			for (int i=0;i<errorCausingPayloads.length;i++){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Could use an enhanced for.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+				msg = getNewMsg();
+				setParameter(msg, param, errorCausingPayloads[i]);
+				sendAndReceive(msg);
+				if(!msg.getResponseBody().toString().contains(errorCausingPayloads[i])){
+					positiveBasedOnError++;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Should break as well, it's not necessary to perform more tests.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+				// Stop the testing if user requests
+				if (isStop()) {
+					if (LOG.isDebugEnabled()){
+						LOG.debug("Stopping the scan due to a user request");
+					}
+					return;
+				}
+			}
+			//vendor returning tests
+			for(int i=0;i<vendorReturningPayloads.length;i++){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Could use an enhanced for.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+				msg=getNewMsg();
+				setParameter(msg,param,vendorReturningPayloads[i]);
+				sendAndReceive(msg);
+				for(int j=0;j<xsltVendors.length;j++){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Could use an enhanced for.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 27]
```diff
@@ -0,0 +1,178 @@
...
+					if(msg.getResponseBody().toString().contains(xsltVendors[j])){
+							if(positiveBasedOnError>0){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: This check should be done earlier, before the `vendorReturningPayloads` for loop, no need to try the tests if there's no error (`positiveBasedOnError` will/does not change its value).

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 28]
```diff
@@ -0,0 +1,178 @@
...
+					if (isStop()) {
+						if (LOG.isDebugEnabled()){
+							LOG.debug("Stopping the scan due to a user request");
+						}
+						return;
+					}
+				}
+			}
+			
+			// TODO add time based testing and improve accuracy
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: I'd move this to the previous TODO.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml [hunk 29]
```diff
@@ -9,7 +9,9 @@
 	<![CDATA[
 	Add .htaccess scanner (Issue 3972).<br>
 	Modified trace.axd scanner to leverage new AbstractAppFilePlugin component.<br>
-	]]>
+	Add XSLT Injection Scanner (Issue 3572).<br>
+	Added ascan rule for xslt injection vulnerabilities.<br>
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml]: This should be removed, with one "change" is enough.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml [hunk 30]
```diff
@@ -9,7 +9,9 @@
 	<![CDATA[
 	Add .htaccess scanner (Issue 3972).<br>
 	Modified trace.axd scanner to leverage new AbstractAppFilePlugin component.<br>
-	]]>
+	Add XSLT Injection Scanner (Issue 3572).<br>
+	Added ascan rule for xslt injection vulnerabilities.<br>
+	]]>
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml]: Trailing tabs added should be removed.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 9 above)
Comment (thc202) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: That can be added as a unit test then.

Comment (thc202) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: > production apps never return errors

I agree with properly configured apps, that's not always the case. Be good to check for errors as well.

**Comment by CaptainFreak** [comment]
Have a look.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html [hunk 31]
```diff
@@ -79,5 +79,8 @@ <H2>Trace.axd Scanner</H2>
 <H2>User Agent Fuzzer</H2>
 This active scanner checks for differences in response based on fuzzed User Agent (eg. mobile sites, access as a Search Engine Crawler). The scanner compares the response statuscode and the hashcode of the response body with the original response.
 
+<H2>XSLT Injection Scanner</H2>
+This active scanner checks for errors and different responses after injecting special xslt character and then finds the context of injection.
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html]: XSLT should probably be full caps, and characters should probably be plural.

**Comment by CaptainFreak** [comment]
Fixed Typos :)

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 32]
```diff
@@ -0,0 +1,162 @@
...
+		try {
+			//Count positive Tests
+			boolean positiveBasedOnError=false; 
+			//error causing tests
+			for (String payload : errorCausingPayloads){
+				msg = getNewMsg();
+				setParameter(msg, param, payload);
+				sendAndReceive(msg);
+				if(!msg.getResponseBody().toString().contains(payload)){
+					positiveBasedOnError=true;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Using your test files under xampp on windows.
This check doesn't seem to be robust enough. Based on the following request `http://localhost:81/xslt/xslt.php?name=%3C&email=me%40my.net` results in:
```html
<me@my.net
<br />
<b>Warning</b>:  DOMDocument::load(): StartTag: invalid element name in file:/C:/kingthorin/ZAP_DEV/xampp/htdocs/xslt/xslt.xsl, line: 3 in <b>C:\kingthorin\ZAP_DEV\xampp\htdocs\xslt\xslt.php</b> on line <b>31</b><br />
<br />
<b>Warning</b>:  XSLTProcessor::importStylesheet(): compilation error in <b>C:\kingthorin\ZAP_DEV\xampp\htdocs\xslt\xslt.php</b> on line <b>35</b><br />
<br />
<b>Warning</b>:  XSLTProcessor::importStylesheet(): xsltParseStylesheetProcess : empty stylesheet in <b>C:\kingthorin\ZAP_DEV\xampp\htdocs\xslt\xslt.php</b> on line <b>35</b><br />
<br />
<b>Warning</b>:  XSLTProcessor::transformToXml(): No stylesheet associated to this object in <b>C:\kingthorin\ZAP_DEV\xampp\htdocs\xslt\xslt.php</b> on line <b>37</b><br />
```
Which obviously contains the injection, but should set `positiveBasedOnError=true`. Yet at the same time obviously web content is going to contain &gt; &lt; and quotes.... I think you're going to have to come up with some error string to check the response for (and will probably likely need to pre-check the original response for those as well, like this: https://github.com/zaproxy/zap-extensions/pull/1342)

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 33]
```diff
@@ -0,0 +1,162 @@
...
+				}
+			}
+			//vendor returning tests
+			if(positiveBasedOnError==true){
+				for(String payload : vendorReturningPayloads){
+					msg=getNewMsg();
+					setParameter(msg,param,payload);
+					sendAndReceive(msg);
+					for(String vendor : xsltVendors){
+						if(msg.getResponseBody().toString().contains(vendor)){
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: There seems to be a bunch of trailing tabs here that can be removed.

Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: The original response should be pre-checked for these or it may lead to false positives.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java [hunk 34]
```diff
@@ -0,0 +1,162 @@
...
+					msg=getNewMsg();
+					setParameter(msg,param,payload);
+					sendAndReceive(msg);
+					for(String vendor : xsltVendors){
+						if(msg.getResponseBody().toString().contains(vendor)){						
+									// We found the injection point
+									bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, getName(), getDescription(),
+									getBaseMsg().getRequestHeader().getURI().toString(),
+									param,  payload,
+									"", getSolution(), getBaseMsg());
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: `getBasemsg()` should be replaced with the message that was actually sent to generate the alert `msg`. 
Actually to include an evidence string the whole bingo should be replaced with:
```
bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, getName(), getDescription(),
		getBaseMsg().getRequestHeader().getURI().toString(), param, payload, "",
		getSolution(), vendor, msg);
```

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml [hunk 35]
```diff
@@ -1,6 +1,6 @@
 <zapaddon>
 	<name>Active scanner rules (alpha)</name>
-	<version>21</version>
+	<version>25</version>
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml]: Manifest version should still be 21, it seems like 20 is the current alpha release:
https://github.com/zaproxy/zap-extensions/releases

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 32 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Note: I did quickly code past this to test further and the second check does return `libxslt`.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 33 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: Like these: https://github.com/zaproxy/zap-extensions/pulls?utf8=%E2%9C%93&q=is%3Apr+pre-check+is%3Aclosed

**Comment by kingthorin** [comment]
@CaptainFreak are you able to address the review comments?

**Comment by CaptainFreak** [comment]
@kingthorin yes. Will completely address them once I get time.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml (hunk 35 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/ZapAddOn.xml]: 21 has been released, you'll have to check when you get to dealing with this and set it properly.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java (hunk 32 above)
Comment (kingthorin) [reply src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java]: @CaptainFreak this is the main hurdle, if you want to tackle this part (or share your ideas). I could probably deal with tweaking the other out standing things.

**Comment by kingthorin** [comment]
@CaptainFreak any news?

**Comment by CaptainFreak** [comment]
Due to my busy schedule will not be able to complete this :(

**Comment by kingthorin** [comment]
@CaptainFreak thanks for letting us know. We'll keep this open and hopefully one of us or someone else can pick it up and finish it.

**Comment by kingthorin** [comment]
Rebased current, fixed up (into a single commit), addressed most feedback.

Still outstanding:
- [ ] https://github.com/zaproxy/zap-extensions/pull/1384#discussion_r173205003 initial check needs to be more robust.

**Comment by psiinon** [comment]
This pull request **introduces 18 alerts** when merging 184c99f11c2e00a130139b019857f92ccbdf74d7 into 27b2116fe841c297bcec0018808d5533344f2f2d - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-a59dd5d39471d6e17e92933f2aac483168445ebd)

**new alerts:**

* 17 for Missing variable declaration
* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by thc202** [comment]
The pull request is now updated (after build/merge changes, zaproxy/zaproxy#5302).

**Comment by thc202** [comment]
Closing, the scanner is now being added/improved in #2170.
